muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/forum/ForumSubscriptionNotifier.java

```diff
@@ -0,0 +1,136 @@
+package fi.otavanopisto.muikku.plugins.forum;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import javax.ejb.AccessTimeout;
+import javax.ejb.Asynchronous;
+import javax.ejb.Singleton;
+import javax.enterprise.event.Observes;
+import javax.enterprise.event.TransactionPhase;
+import javax.inject.Inject;
+import javax.transaction.Transactional;
+import javax.transaction.Transactional.TxType;
+
+import fi.otavanopisto.muikku.model.users.UserEntity;
+import fi.otavanopisto.muikku.notifier.NotifierController;
+import fi.otavanopisto.muikku.plugins.forum.events.ForumMessageSent;
+import fi.otavanopisto.muikku.plugins.forum.model.ForumArea;
+import fi.otavanopisto.muikku.plugins.forum.model.ForumThread;
+import fi.otavanopisto.muikku.plugins.forum.wall.ForumAreaSubscription;
+import fi.otavanopisto.muikku.plugins.forum.wall.ForumThreadSubscription;
+import fi.otavanopisto.muikku.schooldata.entity.User;
+import fi.otavanopisto.muikku.users.UserController;
+import fi.otavanopisto.muikku.users.UserEntityController;
+
+@Singleton
+public class ForumSubscriptionNotifier {
+
+  @Inject
+  private Logger logger;
+ 
+  @Inject
+  private ForumThreadNewMessageNotification forumThreadNewMessageNotification;
+  
+  @Inject
+  private UserEntityController userEntityController;
+  
+  @Inject
+  private UserController userController;
+  
+  @Inject
+  private NotifierController notifierController;
+  
+  @Inject
+  private ForumController forumController;
+  
+  @Inject
+  private ForumThreadSubsciptionController forumThreadSubsciptionController;
+  
+  @Inject
+  private ForumAreaSubsciptionController forumAreaSubsciptionController;
+  
+  @Asynchronous
+  @AccessTimeout(value = 30, unit = TimeUnit.MINUTES)
+  @Transactional (value = TxType.REQUIRES_NEW)
+  public void onForumMessageSent(@Observes (during = TransactionPhase.AFTER_COMPLETION) ForumMessageSent event) {
+    ForumArea forumArea = forumController.getForumArea(event.getAreaId());
+    List<ForumAreaSubscription> areaSubscriptions = null;
+    
+    if (forumArea != null) {
+      areaSubscriptions = forumAreaSubsciptionController.listByArea(forumArea);
+    }
+    
+    ForumThread forumThread = forumController.getForumThread(event.getThreadId());
+    UserEntity poster = userEntityController.findUserEntityById(event.getPosterUserEntityId());
+    
+
+    if (areaSubscriptions != null) {
+      for (ForumAreaSubscription areaSubscription : areaSubscriptions) {
+        Long userId = areaSubscription.getUser();
+        
+        UserEntity recipient = userEntityController.findUserEntityById(userId);
+
+        ForumThreadSubscription forumThreadSubscription = forumThreadSubsciptionController.findByThreadAndUserEntity(forumThread, recipient);
+        
+        if (forumThreadSubscription != null) {
```

Comment:
So if a user has subscribed to both forum area and a forum thread in that area, don't send notification via area, because the user gets notified via thread instead? No biggie but would be nice to add comments so that others can have a better understanding of what is going on.